How do I instruct the go_sdk to use the version specified in `go.mod`?

I have a new dependency in a project, which has a toolchain dependency higher than the local toolchain.

When I run `bazel run //rules_go:go -- mod tidy` I get:

```
bazel run @rules_go//go -- mod tidy
Computing main repo mapping: 
Loading: 
Loading: 0 packages loaded
bazel: Entering directory `/home/jaq/.cache/bazel/_bazel_jaq/d654f9f0bd3eddeb1a4864cf68fcdd6b/execroot/_main/'
Analyzing: target @@rules_go+//go:go (0 packages loaded, 0 targets configured)
Analyzing: target @@rules_go+//go:go (0 packages loaded, 0 targets configured)

INFO: Analyzed target @@rules_go+//go:go (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
bazel: Leaving directory `/home/jaq/.cache/bazel/_bazel_jaq/d654f9f0bd3eddeb1a4864cf68fcdd6b/execroot/_main/'
Target @@rules_go+//go/tools/go_bin_runner:go_bin_runner up-to-date:
  bazel-bin/external/rules_go+/go/tools/go_bin_runner/bin/go
INFO: Elapsed time: 0.173s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/external/rules_go+/go/tools/go_bin_runner/bin/go <args omitted>
go: finding module for package github.com/google/go-containerregistry/pkg/v1/layout
go: toolchain upgrade needed to resolve github.com/google/go-containerregistry/pkg/v1/layout
go: github.com/google/go-containerregistry@v0.20.3 requires go >= 1.23.0 (running go 1.22.7; GOTOOLCHAIN=local)

Compilation exited abnormally with code 1 at Thu Mar 13 20:10:26, duration 3.03 s
```

The go.mod file contains this at the top:

```
module github.com/jaqx0r/blts

go 1.22

toolchain go1.24.1

```

The hint is `GOTOOLCHAIN=local` in the error.

I haven't set an explicit toolchain in `MODULE.bazel` because I assumed it would figure it out from `go.mod`:

```
module(name = "blts")

bazel_dep(name = "rules_go", version = "0.53.0")
bazel_dep(name = "gazelle", version = "0.42.0")

go_sdk = use_extension("@rules_go//go:extensions.bzl", "go_sdk")

# https://github.com/bazel-contrib/rules_go/blob/master/go/nogo.rst
# Like golangci-lint, but integrated into the build.
go_sdk.nogo(nogo = "//:nogo")

# Update dependencies with
# `bazel run //:gazelle`.
# and then
# `bazel run @rules_go//go -- mod tidy`
go_deps = use_extension("@gazelle//:extensions.bzl", "go_deps")
go_deps.from_file(go_mod = "//:go.mod")
use_repo(go_deps, ...)
```

I _expected_ the default behaviour to be hermetic, that is I would not be using the system Go SDK.  BUt even that doesn't make sense:

```
% go version
go version go1.24.1 linux/amd64
```

so wherever this `running go 1.22.7; GOTOOLCHAIN=local` is coming from, I don't know.

```
slab% go env GOTOOLCHAIN 
auto
```

The go version is supposed to be defined in `go.mod` so how can I make the `go_sdk.download` call in `MODULE.bazel` read this from `go.mod` so I don't have to repeat myself?

Thanks @fmeum.  I'm still missing something though, because now it looks like I correctly set the `first_host_compatible_toolchain` to be my toolchain selected from my `from_file` tag, but it isn't the one being used to build and run the test code, which is the same problem in the output above.

I've uploaded a new commit, with some extra debug statements printing out state. 

https://github.com/bazel-contrib/rules_go/compare/master...jaqx0r:rules_go:from-go-mod-file

with the test output:


```
bazel test //tests/core/from_go_mod_file/...
Computing main repo mapping: 
Loading: 
Loading: 0 packages loaded
bazel: Entering directory `/home/jaq/.cache/bazel/_bazel_jaq/906827e386e14cc47d13c9784c62deb4/execroot/io_bazel_rules_go/'
Analyzing: target //tests/core/from_go_mod_file:from_go_mod_file_test (0 packages loaded, 0 targets configured)
Analyzing: target //tests/core/from_go_mod_file:from_go_mod_file_test (0 packages loaded, 0 targets configured)

INFO: Analyzed target //tests/core/from_go_mod_file:from_go_mod_file_test (0 packages loaded, 0 targets configured).
[10 / 11] Testing //tests/core/from_go_mod_file:from_go_mod_file_test; 1s local
FAIL: //tests/core/from_go_mod_file:from_go_mod_file_test (see /home/jaq/.cache/bazel/_bazel_jaq/906827e386e14cc47d13c9784c62deb4/execroot/io_bazel_rules_go/bazel-out/k8-fastbuild/testlogs/tests/core/from_go_mod_file/from_go_mod_file_test/test.log)
[10 / 11] 1 / 1 tests, 1 failed; Testing //tests/core/from_go_mod_file:from_go_mod_file_test; 8s local
INFO: From Testing //tests/core/from_go_mod_file:from_go_mod_file_test:
==================== Test output for //tests/core/from_go_mod_file:from_go_mod_file_test:
--- FAIL: Test (7.43s)
    --- FAIL: Test/toolchain (7.43s)
        from_go_mod_file_test.go:97: Starting local Bazel server and connecting to it...
            Computing main repo mapping: 
            Computing main repo mapping: 
            Computing main repo mapping: 
            Computing main repo mapping: 
            Loading: 
            Loading: 0 packages loaded
            DEBUG: /home/jaq/.cache/bazel/_bazel_jaq/7a087131b1e4a89e6bfe21cd9cecf749/external/rules_go~/go/private/extensions.bzl:176:18: Got version 1.24.1
            DEBUG: /home/jaq/.cache/bazel/_bazel_jaq/7a087131b1e4a89e6bfe21cd9cecf749/external/rules_go~/go/private/extensions.bzl:323:10: toolchains: [struct(goarch = "", goos = "", sdk_repo = "main___from_file_0", sdk_type = "remote", sdk_version = "1.24.1"), struct(goarch = "", goos = "", sdk_repo = "go_default_sdk", sdk_type = "remote", sdk_version = "1.23.6"), struct(goarch = "amd64", goos = "darwin", sdk_repo = "rules_go__download_0_darwin_amd64", sdk_type = "remote", sdk_version = "1.23.6"), struct(goarch = "arm64", goos = "darwin", sdk_repo = "rules_go__download_0_darwin_arm64", sdk_type = "remote", sdk_version = "1.23.6"), struct(goarch = "arm64", goos = "linux", sdk_repo = "rules_go__download_0_linux_arm64", sdk_type = "remote", sdk_version = "1.23.6"), struct(goarch = "amd64", goos = "windows", sdk_repo = "rules_go__download_0_windows_amd64", sdk_type = "remote", sdk_version = "1.23.6"), struct(goarch = "arm64", goos = "windows", sdk_repo = "rules_go__download_0_windows_arm64", sdk_type = "remote", sdk_version = "1.23.6")]
            DEBUG: /home/jaq/.cache/bazel/_bazel_jaq/7a087131b1e4a89e6bfe21cd9cecf749/external/rules_go~/go/private/extensions.bzl:324:10: first : @main___from_file_0//:ROOT
            Analyzing: target //:version_test (1 packages loaded, 0 targets configured)
            Analyzing: target //:version_test (1 packages loaded, 0 targets configured)
            
            Analyzing: target //:version_test (91 packages loaded, 918 targets configured)
            
            INFO: Analyzed target //:version_test (107 packages loaded, 5383 targets configured).
            FAIL: //:version_test (see /home/jaq/.cache/bazel/_bazel_jaq/7a087131b1e4a89e6bfe21cd9cecf749/execroot/_main/bazel-out/k8-fastbuild/testlogs/version_test/test.log)
            INFO: From Testing //:version_test:
            INFO: Found 1 test target...
            Target //:version_test up-to-date:
              bazel-bin/version_test_/version_test
            INFO: Elapsed time: 7.335s, Critical Path: 0.15s
            INFO: 3 processes: 2 internal, 1 linux-sandbox.
            INFO: Build completed, 1 test FAILED, 3 total actions
            exit status 3
FAIL
```

and the content of the inner test's log:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //:version_test
-----------------------------------------------------------------------------
go1.24.0 X:nocoverageredesign
--- FAIL: Test (0.00s)
    version_test.go:15: got version "go1.24.0 X:nocoverageredesign"; want "go1.24.1"
FAIL
```

So you can see in extensions.bzl:176 we've correctly read the toolchain 1.24.1 from the go.mod in test.

The full set of toolchains then printed out has our `from_file` toolchain first, and the `first_host_compatible_toolchain` has the correct `sdk_repo` prefix.

But when the inner test executes, it's runtime is reported as `1.24.0`, not `1.24.1`.   Have I messed up the test?

--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -16,9 +16,9 @@ bazel_dep(name = "protobuf", version = "3.19.2", repo_name = "com_google_protobu
 bazel_dep(name = "rules_shell", version = "0.3.0")
 
 go_sdk = use_extension("//go:extensions.bzl", "go_sdk")
-go_sdk.download(
+go_sdk.from_file(
     name = "go_default_sdk",
-    version = "1.23.6",
+    go_mod = "//:go.mod",
 )
 use_repo(
     go_sdk,

--- a/docs/go/core/bzlmod.md
+++ b/docs/go/core/bzlmod.md
@@ -30,7 +30,12 @@ To register a particular version of the Go SDK, use the `go_sdk` module extensio
 ```starlark
 go_sdk = use_extension("@rules_go//go:extensions.bzl", "go_sdk")
 
-# Download an SDK for the host OS & architecture as well as common remote execution platforms.
+# Download an SDK for the host OS & architecture as well as common remote execution
+# platforms, using the version given from the `go.mod` file.
+go_sdk.from_file(go_mod = "//:go.mod")
+
+# Download an SDK for the host OS & architecture as well as common remote execution
+# platforms, with a specific version.
 go_sdk.download(version = "1.23.1")
 
 # Alternatively, download an SDK for a fixed OS/architecture.

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -215,7 +215,7 @@ def _go_sdk_impl(ctx):
                 sdk_version = wrap_tag.version,
             ))
 
-        # First if the module suggests to read the toolchain version from a `go.mod` file, use that.
+        # If the module suggests to read the toolchain version from a `go.mod` file, use that.
         for index, from_file_tag in enumerate(module.tags.from_file):
             version = version_from_go_mod(ctx, from_file_tag.go_mod)
             name = from_file_tag.name or _default_go_sdk_name(

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -56,6 +56,7 @@ _COMMON_TAG_ATTRS = {
 }
 
 _download_tag = tag_class(
+    doc = """Download a specific Go SDK at the optional GOOS, GOARCH, and version, from a customisable URL.  Optionally apply local customisations to the SDK though patches and experiments.""",
     attrs = _COMMON_TAG_ATTRS | {
         "version": attr.string(),
     },
@@ -117,6 +118,7 @@ _wrap_tag = tag_class(
 )
 
 _from_file_tag = tag_class(
+    doc = """Use a specific Go SDK version described by a `go.mod` file.  Optionally supply GOOS, GOARCH, and download from a customisable URL, with any local patches or experiments applied.""",
     attrs = _COMMON_TAG_ATTRS | {
         "go_mod": attr.label(
             doc = "The go.mod file to read the SDK version from.",

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -36,25 +36,28 @@ host_compatible_toolchain = repository_rule(
     doc = "An external repository to expose the first host compatible toolchain",
 )
 
+_COMMON_TAG_ATTRS = {
+    "name": attr.string(),
+    "goos": attr.string(),
+    "goarch": attr.string(),
+    "sdks": attr.string_list_dict(),
+    "experiments": attr.string_list(
+        doc = "Go experiments to enable via GOEXPERIMENT",
+    ),
+    "urls": attr.string_list(default = ["https://dl.google.com/go/{}"]),
+    "patches": attr.label_list(
+        doc = "A list of patches to apply to the SDK after downloading it",
+    ),
+    "patch_strip": attr.int(
+        default = 0,
+        doc = "The number of leading path segments to be stripped from the file name in the patches.",
+    ),
+    "strip_prefix": attr.string(default = "go"),
+}
+
 _download_tag = tag_class(
-    attrs = {
-        "name": attr.string(),
-        "goos": attr.string(),
-        "goarch": attr.string(),
-        "sdks": attr.string_list_dict(),
-        "experiments": attr.string_list(
-            doc = "Go experiments to enable via GOEXPERIMENT",
-        ),
-        "urls": attr.string_list(default = ["https://dl.google.com/go/{}"]),
+    attrs = _COMMON_TAG_ATTRS | {
         "version": attr.string(),
-        "patches": attr.label_list(
-            doc = "A list of patches to apply to the SDK after downloading it",
-        ),
-        "patch_strip": attr.int(
-            default = 0,
-            doc = "The number of leading path segments to be stripped from the file name in the patches.",
-        ),
-        "strip_prefix": attr.string(default = "go"),
     },
 )
 
@@ -114,8 +117,7 @@ _wrap_tag = tag_class(
 )
 
 _from_file_tag = tag_class(
-    attrs = {
-        "name": attr.string(),
+    attrs = _COMMON_TAG_ATTRS | {
         "go_mod": attr.label(
             doc = "The go.mod file to read the SDK version from.",
         ),
@@ -210,6 +212,41 @@ def _go_sdk_impl(ctx):
                 sdk_type = "remote",
                 sdk_version = wrap_tag.version,
             ))
+
+        # First if the module suggests to read the toolchain version from a `go.mod` file, use that.
+        for index, from_file_tag in enumerate(module.tags.from_file):
+            version = version_from_go_mod(ctx, from_file_tag.go_mod)
+            print("Got version {}".format(version))
+            name = from_file_tag.name or _default_go_sdk_name(
+                module = module,
+                multi_version = multi_version_module[module.name],
+                tag_type = "from_file",
+                index = index,
+            )
+
+            # Keep in sync with the other calls to `go_download_sdk` below.
+            go_download_sdk_rule(
+                name = name,
+                version = version,
+                goos = from_file_tag.goos,
+                goarch = from_file_tag.goarch,
+                sdks = from_file_tag.sdks,
+                experiments = from_file_tag.experiments,
+                patches = from_file_tag.patches,
+                patch_strip = from_file_tag.patch_strip,
+                urls = from_file_tag.urls,
+                strip_prefix = from_file_tag.strip_prefix,
+            )
+            if (not from_file_tag.goos or from_file_tag.goos == host_detected_goos) and (not from_file_tag.goarch or from_file_tag.goarch == host_detected_goarch):
+                first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
+            toolchains.append(struct(
+                goos = from_file_tag.goos,
+                goarch = from_file_tag.goarch,
+                sdk_repo = name,
+                sdk_type = "remote",
+                sdk_version = version,
+            ))
+            first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
 
         for index, download_tag in enumerate(module.tags.download):
             # SDKs without an explicit version are fetched even when not selected by toolchain
@@ -326,29 +363,8 @@ def _go_sdk_impl(ctx):
             ))
             first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
 
-        for index, from_file_tag in enumerate(module.tags.from_file):
-            version = version_from_go_mod(ctx, from_file_tag.go_mod)
-            print("Got version {}".format(version))
-            name = from_file_tag.name or _default_go_sdk_name(
-                module = module,
-                multi_version = multi_version_module[module.name],
-                tag_type = "from_file",
-                index = index,
-            )
-            go_download_sdk_rule(
-                name = name,
-                version = version,
-            )
-            toolchains.append(struct(
-                goos = "",
-                goarch = "",
-                sdk_repo = name,
-                sdk_type = "remote",
-                sdk_version = version,
-            ))
-            first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
-
     print("toolchains: {}".format(toolchains))
+    print("first : {}".format(first_host_compatible_toolchain))
 
     host_compatible_toolchain(name = "go_host_compatible_sdk_label", toolchain = first_host_compatible_toolchain)
     if len(toolchains) > _MAX_NUM_TOOLCHAINS:

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -218,7 +218,6 @@ def _go_sdk_impl(ctx):
         # First if the module suggests to read the toolchain version from a `go.mod` file, use that.
         for index, from_file_tag in enumerate(module.tags.from_file):
             version = version_from_go_mod(ctx, from_file_tag.go_mod)
-            print("Got version {}".format(version))
             name = from_file_tag.name or _default_go_sdk_name(
                 module = module,
                 multi_version = multi_version_module[module.name],
@@ -364,9 +363,6 @@ def _go_sdk_impl(ctx):
                 sdk_version = host_tag.version,
             ))
             first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
-
-    print("toolchains: {}".format(toolchains))
-    print("first : {}".format(first_host_compatible_toolchain))
 
     host_compatible_toolchain(name = "go_host_compatible_sdk_label", toolchain = first_host_compatible_toolchain)
     if len(toolchains) > _MAX_NUM_TOOLCHAINS:

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 load("@io_bazel_rules_go_bazel_features//:features.bzl", "bazel_features")
+load("//go/private:go_mod.bzl", "version_from_go_mod")
 load("//go/private:nogo.bzl", "DEFAULT_NOGO", "NOGO_DEFAULT_EXCLUDES", "NOGO_DEFAULT_INCLUDES", "go_register_nogo")
 load("//go/private:sdk.bzl", "detect_host_platform", "go_download_sdk_rule", "go_host_sdk_rule", "go_multiple_toolchains", "go_wrap_sdk_rule")
 
@@ -109,6 +110,15 @@ _wrap_tag = tag_class(
         ),
         "goos": attr.string(),
         "goarch": attr.string(),
+    },
+)
+
+_from_file_tag = tag_class(
+    attrs = {
+        "name": attr.string(),
+        "go_mod": attr.label(
+            doc = "The go.mod file to read the SDK version from.",
+        ),
     },
 )
 
@@ -316,6 +326,30 @@ def _go_sdk_impl(ctx):
             ))
             first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
 
+        for index, from_file_tag in enumerate(module.tags.from_file):
+            version = version_from_go_mod(ctx, from_file_tag.go_mod)
+            print("Got version {}".format(version))
+            name = from_file_tag.name or _default_go_sdk_name(
+                module = module,
+                multi_version = multi_version_module[module.name],
+                tag_type = "from_file",
+                index = index,
+            )
+            go_download_sdk_rule(
+                name = name,
+                version = version,
+            )
+            toolchains.append(struct(
+                goos = "",
+                goarch = "",
+                sdk_repo = name,
+                sdk_type = "remote",
+                sdk_version = version,
+            ))
+            first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
+
+    print("toolchains: {}".format(toolchains))
+
     host_compatible_toolchain(name = "go_host_compatible_sdk_label", toolchain = first_host_compatible_toolchain)
     if len(toolchains) > _MAX_NUM_TOOLCHAINS:
         fail("more than {} go_sdk tags are not supported".format(_MAX_NUM_TOOLCHAINS))
@@ -383,6 +417,7 @@ go_sdk = module_extension(
         "host": _host_tag,
         "nogo": _nogo_tag,
         "wrap": _wrap_tag,
+        "from_file": _from_file_tag,
     },
     **go_sdk_extra_kwargs
 )

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -56,7 +56,7 @@ _COMMON_TAG_ATTRS = {
 }
 
 _download_tag = tag_class(
-    doc = """Download a specific Go SDK at the optional GOOS, GOARCH, and version, from a customisable URL.  Optionally apply local customisations to the SDK though patches and experiments.""",
+    doc = """Download a specific Go SDK at the optional GOOS, GOARCH, and version, from a customisable URL.  Optionally apply local customisations to the SDK by applying patches and setting experiments.""",
     attrs = _COMMON_TAG_ATTRS | {
         "version": attr.string(),
     },
@@ -118,7 +118,7 @@ _wrap_tag = tag_class(
 )
 
 _from_file_tag = tag_class(
-    doc = """Use a specific Go SDK version described by a `go.mod` file.  Optionally supply GOOS, GOARCH, and download from a customisable URL, with any local patches or experiments applied.""",
+    doc = """Use a specific Go SDK version described by a `go.mod` file.  Optionally supply GOOS, GOARCH, and download from a customisable URL, and apply local patches or set experiments.""",
     attrs = _COMMON_TAG_ATTRS | {
         "go_mod": attr.label(
             doc = "The go.mod file to read the SDK version from.",
@@ -271,7 +271,7 @@ def _go_sdk_impl(ctx):
                 index = index,
             )
 
-            # Keep in sync with the other call to go_download_sdk_rule below.
+            # Keep in sync with the other calls to `go_download_sdk_rule` above and below.
             go_download_sdk_rule(
                 name = name,
                 goos = download_tag.goos,
@@ -315,6 +315,8 @@ def _go_sdk_impl(ctx):
                         index = index,
                         suffix = "_{}_{}".format(goos, goarch),
                     )
+
+                    # Keep in sync with the other calls to `go_download_sdk_rule` above.
                     go_download_sdk_rule(
                         name = default_name,
                         goos = goos,

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -215,41 +215,22 @@ def _go_sdk_impl(ctx):
                 sdk_version = wrap_tag.version,
             ))
 
+        additional_download_tags = []
+
         # If the module suggests to read the toolchain version from a `go.mod` file, use that.
         for index, from_file_tag in enumerate(module.tags.from_file):
             version = version_from_go_mod(ctx, from_file_tag.go_mod)
-            name = from_file_tag.name or _default_go_sdk_name(
-                module = module,
-                multi_version = multi_version_module[module.name],
-                tag_type = "from_file",
-                index = index,
-            )
 
-            # Keep in sync with the other calls to `go_download_sdk` below.
-            go_download_sdk_rule(
-                name = name,
-                version = version,
-                goos = from_file_tag.goos,
-                goarch = from_file_tag.goarch,
-                sdks = from_file_tag.sdks,
-                experiments = from_file_tag.experiments,
-                patches = from_file_tag.patches,
-                patch_strip = from_file_tag.patch_strip,
-                urls = from_file_tag.urls,
-                strip_prefix = from_file_tag.strip_prefix,
-            )
-            if (not from_file_tag.goos or from_file_tag.goos == host_detected_goos) and (not from_file_tag.goarch or from_file_tag.goarch == host_detected_goarch):
-                first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
-            toolchains.append(struct(
-                goos = from_file_tag.goos,
-                goarch = from_file_tag.goarch,
-                sdk_repo = name,
-                sdk_type = "remote",
-                sdk_version = version,
-            ))
-            first_host_compatible_toolchain = first_host_compatible_toolchain or "@{}//:ROOT".format(name)
+            # Synthesize a `download` tag so we can reuse the selection logic below.
+            download_tag = {
+                key: getattr(from_file_tag, key)
+                for key in dir(from_file_tag)
+                if key not in "go_mod"
+            }
+            download_tag["version"] = version
+            additional_download_tags += [struct(**download_tag)]
 
-        for index, download_tag in enumerate(module.tags.download):
+        for index, download_tag in enumerate(module.tags.download + additional_download_tags):
             # SDKs without an explicit version are fetched even when not selected by toolchain
             # resolution. This is acceptable if brought in by the root module, but transitive
             # dependencies should not slow down the build in this way.

--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -225,7 +225,7 @@ def _go_sdk_impl(ctx):
             download_tag = {
                 key: getattr(from_file_tag, key)
                 for key in dir(from_file_tag)
-                if key not in "go_mod"
+                if key not in ["go_mod"]
             }
             download_tag["version"] = version
             additional_download_tags += [struct(**download_tag)]

--- a/go/private/go_mod.bzl
+++ b/go/private/go_mod.bzl
@@ -1,0 +1,137 @@
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def version_from_go_mod(module_ctx, go_mod_label):
+    """Returns a version string from a go.mod file.
+
+    Args:
+       module_ctx: a https://bazel.build/rules/lib/module_ctx object passed
+            from the MODULE.bazel call.
+       go_mod_label: a Label for a `go.mod` file.
+
+    Returns:
+      a string containing the version of the Go SDK defined in go.mod
+    """
+    _check_go_mod_name(go_mod_label.name)
+    go_mod_path = module_ctx.path(go_mod_label)
+    go_mod_content = module_ctx.read(go_mod_path)
+
+    state = {
+        "toolchain": None,
+        "go": None,
+    }
+
+    for line_no, line in enumerate(go_mod_content.splitlines(), 1):
+        tokens, _ = _tokenize_line(line, go_mod_path, line_no)
+        if not tokens:
+            continue
+
+        if tokens[0] not in ["module", "go", "require", "replace", "exclude", "retract", "toolchain"]:
+            fail("{}:{}: unexpected token '{}' at start of line".format(go_mod_path, line_no, tokens[0]))
+        if len(tokens) == 1:
+            fail("{}:{}: expected another token after '{}'".format(go_mod_path, line_no, tokens[0]))
+
+        if tokens[0] == "go":
+            _validate_go_version(go_mod_path, state, tokens, line_no)
+            state["go"] = tokens[1]
+
+        if tokens[0] == "toolchain":
+            _validate_toolchain_version(go_mod_path, state, tokens, line_no)
+            state["toolchain"] = tokens[1][len("go"):].strip()
+
+    version = state["toolchain"]
+    if not version:
+        version = state["go"]
+    if not version:
+        # "As of the Go 1.17 release, if the go directive is missing, go 1.16 is assumed."
+        version = "1.16"
+
+    return version
+
+def _tokenize_line(line, path, line_no):
+    tokens = []
+    r = line
+    for _ in range(len(line)):
+        r = r.strip()
+        if not r:
+            break
+
+        if r[0] == "`":
+            end = r.find("`", 1)
+            if end == -1:
+                fail("{}:{}: unterminated raw string".format(path, line_no))
+
+            tokens.append(r[1:end])
+            r = r[end + 1:]
+
+        elif r[0] == "\"":
+            value = ""
+            escaped = False
+            found_end = False
+            pos = 0
+            for pos in range(1, len(r)):
+                c = r[pos]
+
+                if escaped:
+                    value += c
+                    escaped = False
+                    continue
+
+                if c == "\\":
+                    escaped = True
+                    continue
+
+                if c == "\"":
+                    found_end = True
+                    break
+
+                value += c
+
+            if not found_end:
+                fail("{}:{}: unterminated interpreted string".format(path, line_no))
+
+            tokens.append(value)
+            r = r[pos + 1:]
+
+        elif r.startswith("//"):
+            # A comment always ends the current line
+            return tokens, r[len("//"):].strip()
+
+        else:
+            token, _, r = r.partition(" ")
+            tokens.append(token)
+
+    return tokens, None
+
+def _check_go_mod_name(name):
+    if name != "go.mod":
+        fail("go_sdk.from_file requires a 'go.mod' file, not '{}'".format(name))
+
+def _validate_go_version(path, state, tokens, line_no):
+    if len(tokens) == 1:
+        fail("{}:{}: expected another token after 'go'".format(path, line_no))
+    if state["go"] != None:
+        fail("{}:{}: unexpected second 'go' directive".format(path, line_no))
+    if len(tokens) > 2:
+        fail("{}:{}: unexpected token '{}' after '{}'".format(path, line_no, tokens[2], tokens[1]))
+
+def _validate_toolchain_version(path, state, tokens, line_no):
+    if len(tokens) == 1:
+        fail("{}:{}: expected another token after 'toolchain'".format(path, line_no))
+    if state["toolchain"] != None:
+        fail("{}:{}: unexpected second 'toolchain' directive".format(path, line_no))
+    if len(tokens) > 2:
+        fail("{}:{}: unexpected token '{}' after '{}'".format(path, line_no, tokens[2], tokens[1]))
+    if not tokens[1].startswith("go"):
+        fail("{}:{}: expected toolchain version to start with 'go', not '{}'".format(path, line_no, tokens[1]))

--- a/go/private/go_mod.bzl
+++ b/go/private/go_mod.bzl
@@ -32,23 +32,36 @@ def version_from_go_mod(module_ctx, go_mod_label):
         "go": None,
     }
 
+    current_directive = None
     for line_no, line in enumerate(go_mod_content.splitlines(), 1):
         tokens, _ = _tokenize_line(line, go_mod_path, line_no)
         if not tokens:
             continue
 
-        if tokens[0] not in ["module", "go", "require", "replace", "exclude", "retract", "toolchain"]:
-            fail("{}:{}: unexpected token '{}' at start of line".format(go_mod_path, line_no, tokens[0]))
-        if len(tokens) == 1:
-            fail("{}:{}: expected another token after '{}'".format(go_mod_path, line_no, tokens[0]))
+        if not current_directive:
+            if tokens[0] not in ["module", "go", "require", "replace", "exclude", "retract", "toolchain"]:
+                fail("{}:{}: unexpected token '{}' at start of line".format(go_mod_path, line_no, tokens[0]))
+            if len(tokens) == 1:
+                fail("{}:{}: expected another token after '{}'".format(go_mod_path, line_no, tokens[0]))
 
-        if tokens[0] == "go":
-            _validate_go_version(go_mod_path, state, tokens, line_no)
-            state["go"] = tokens[1]
+            if tokens[0] == "go":
+                _validate_go_version(go_mod_path, state, tokens, line_no)
+                state["go"] = tokens[1]
 
-        if tokens[0] == "toolchain":
-            _validate_toolchain_version(go_mod_path, state, tokens, line_no)
-            state["toolchain"] = tokens[1][len("go"):].strip()
+            if tokens[0] == "toolchain":
+                _validate_toolchain_version(go_mod_path, state, tokens, line_no)
+                state["toolchain"] = tokens[1][len("go"):].strip()
+
+            if tokens[1] == "(":
+                current_directive = tokens[0]
+                if len(tokens) > 2:
+                    fail("{}:{}: unexpected token '{}' after '('".format(go_mod_path, line_no, tokens[2]))
+                continue
+        elif tokens[0] == ")":
+            current_directive = None
+            if len(tokens) > 1:
+                fail("{}:{}: unexpected token '{}' after ')'".format(go_mod_path, line_no, tokens[1]))
+            continue
 
     version = state["toolchain"]
     if not version:

--- a/go/private/go_mod.bzl
+++ b/go/private/go_mod.bzl
@@ -39,11 +39,6 @@ def version_from_go_mod(module_ctx, go_mod_label):
             continue
 
         if not current_directive:
-            if tokens[0] not in ["module", "go", "require", "replace", "exclude", "retract", "toolchain"]:
-                fail("{}:{}: unexpected token '{}' at start of line".format(go_mod_path, line_no, tokens[0]))
-            if len(tokens) == 1:
-                fail("{}:{}: expected another token after '{}'".format(go_mod_path, line_no, tokens[0]))
-
             if tokens[0] == "go":
                 _validate_go_version(go_mod_path, state, tokens, line_no)
                 state["go"] = tokens[1]

--- a/go/private/go_mod.bzl
+++ b/go/private/go_mod.bzl
@@ -65,6 +65,7 @@ def version_from_go_mod(module_ctx, go_mod_label):
 
     version = state["toolchain"]
     if not version:
+        # https://go.dev/doc/toolchain: "a go.mod that says go 1.21.0 with no toolchain line is interpreted as if it had a toolchain go1.21.0 line."
         version = state["go"]
     if not version:
         # "As of the Go 1.17 release, if the go directive is missing, go 1.16 is assumed."

--- a/tests/core/from_go_mod_file/BUILD.bazel
+++ b/tests/core/from_go_mod_file/BUILD.bazel
@@ -1,0 +1,6 @@
+load("@io_bazel_rules_go//go/tools/bazel_testing:def.bzl", "go_bazel_test")
+
+go_bazel_test(
+    name = "from_go_mod_file_test",
+    srcs = ["from_go_mod_file_test.go"],
+)

--- a/tests/core/from_go_mod_file/README.rst
+++ b/tests/core/from_go_mod_file/README.rst
@@ -1,0 +1,7 @@
+from_go_mod_file
+================
+
+from_go_mod_file_test
+---------------------
+Verifies that ``go_sdk.from_file`` can be used to fetch the Go SDK version
+that is described by the toolchain in ``go.mod``.

--- a/tests/core/from_go_mod_file/from_go_mod_file_test.go
+++ b/tests/core/from_go_mod_file/from_go_mod_file_test.go
@@ -53,7 +53,7 @@ func Test(t *testing.T) {
 `,
 		ModuleFileSuffix: `
 go_sdk = use_extension("@io_bazel_rules_go//go:extensions.bzl", "go_sdk")
-go_sdk.from_file(name = "go_sdk", go_mod = "//:go.mod")
+go_sdk.from_file(go_mod = "//:go.mod")
 `,
 	})
 }
@@ -73,15 +73,15 @@ toolchain go1.24.1
 `,
 			want: "go1.24.1",
 		},
-		{
-			desc: "go only",
-			go_mod: `
-module test
+// 		{
+// 			desc: "go only",
+// 			go_mod: `
+// module test
 
-go 1.23.0
-`,
-			want: "go1.23.0",
-		},
+// go 1.23.0
+// `,
+// 			want: "go1.23.0",
+// 		},
 	} {
 		t.Run(test.desc, func(t *testing.T) {
 			if err := ioutil.WriteFile("go.mod", []byte(test.go_mod), 0o666); err != nil {

--- a/tests/core/from_go_mod_file/from_go_mod_file_test.go
+++ b/tests/core/from_go_mod_file/from_go_mod_file_test.go
@@ -62,49 +62,49 @@ func Test(t *testing.T) {
 	for _, test := range []struct{
 		desc, go_mod, want string
 	}{
-// 		{
-// 			desc: "toolchain",
-// 			go_mod: `
-// module test
+		{
+			desc: "toolchain",
+			go_mod: `
+module test
 
-// go 1.23.0
+go 1.23.0
 
-// toolchain go1.24.0
+toolchain go1.24.0
 
-// require (
-//     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
-// )
-// `,
-// 			want: "go1.24.0 X:nocoverageredesign",
-// 		},
-// 		{
-// 			desc: "toolchain minor version",
-// 			go_mod: `
-// module test
+require (
+    github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+)
+`,
+			want: "go1.24.0 X:nocoverageredesign",
+		},
+		{
+			desc: "toolchain minor version",
+			go_mod: `
+module test
 
-// go 1.23.0
+go 1.23.0
 
-// toolchain go1.24.1
+toolchain go1.24.1
 
-// require (
-//     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
-// )
-// `,
-// 			want: "go1.24.1 X:nocoverageredesign",
-// 		},
-// 		{
-// 			desc: "go only",
-// 			go_mod: `
-// module test
+require (
+    github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+)
+`,
+			want: "go1.24.1 X:nocoverageredesign",
+		},
+		{
+			desc: "go only",
+			go_mod: `
+module test
 
-// go 1.17
+go 1.17
 
-// require (
-//     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
-// )
-// `,
-// 			want: "go1.23.0",
-// 		},
+require (
+    github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+)
+`,
+			want: "go1.23.0",
+		},
 		{
 			desc: "missing go",
 			go_mod: `
@@ -114,7 +114,7 @@ require (
     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
 )
 `,
-			want: "go1.17",
+			want: "go1.16",
 		},
 	} {
 		t.Run(test.desc, func(t *testing.T) {
@@ -126,6 +126,11 @@ require (
 				"--enable_bzlmod",
 				"--test_arg=-version=" + test.want,
 				"--test_output=all",
+				"--sandbox_debug",
+				// This feels like a hack to force the test environment to
+				// choose our SDK, because `bazel_testing` uses `go_wrap_sdk`
+				// to create its own SDK.
+				"--@io_bazel_rules_go//go/toolchain:sdk_name=sdk_under_test",
 				"//:version_test",
 			}
 			if err := bazel_testing.RunBazel(args...); err != nil {

--- a/tests/core/from_go_mod_file/from_go_mod_file_test.go
+++ b/tests/core/from_go_mod_file/from_go_mod_file_test.go
@@ -97,13 +97,13 @@ require (
 			go_mod: `
 module test
 
-go 1.17
+go 1.16
 
 require (
     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
 )
 `,
-			want: "go1.23.0",
+			want: "go1.16",
 		},
 		{
 			desc: "missing go",
@@ -123,13 +123,10 @@ require (
 			}
 			args := []string{
 				"test",
-				"--enable_bzlmod",
 				"--test_arg=-version=" + test.want,
-				"--test_output=all",
-				"--sandbox_debug",
-				// This feels like a hack to force the test environment to
-				// choose our SDK, because `bazel_testing` uses `go_wrap_sdk`
-				// to create its own SDK.
+				// This next flag forces the test environment to choose its own
+				// module's SDK, because `bazel_testing` uses `go_wrap_sdk` to
+				// create its own SDK in the WORKSPACE file.
 				"--@io_bazel_rules_go//go/toolchain:sdk_name=sdk_under_test",
 				"//:version_test",
 			}

--- a/tests/core/from_go_mod_file/from_go_mod_file_test.go
+++ b/tests/core/from_go_mod_file/from_go_mod_file_test.go
@@ -53,7 +53,7 @@ func Test(t *testing.T) {
 `,
 		ModuleFileSuffix: `
 go_sdk = use_extension("@io_bazel_rules_go//go:extensions.bzl", "go_sdk")
-go_sdk.from_file(go_mod = "//:go.mod")
+go_sdk.from_file(name = "sdk_under_test", go_mod = "//:go.mod")
 `,
 	})
 }
@@ -62,26 +62,60 @@ func Test(t *testing.T) {
 	for _, test := range []struct{
 		desc, go_mod, want string
 	}{
-		{
-			desc: "toolchain",
-			go_mod: `
-module test
+// 		{
+// 			desc: "toolchain",
+// 			go_mod: `
+// module test
 
-go 1.23.0
+// go 1.23.0
 
-toolchain go1.24.1
-`,
-			want: "go1.24.1",
-		},
+// toolchain go1.24.0
+
+// require (
+//     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+// )
+// `,
+// 			want: "go1.24.0 X:nocoverageredesign",
+// 		},
+// 		{
+// 			desc: "toolchain minor version",
+// 			go_mod: `
+// module test
+
+// go 1.23.0
+
+// toolchain go1.24.1
+
+// require (
+//     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+// )
+// `,
+// 			want: "go1.24.1 X:nocoverageredesign",
+// 		},
 // 		{
 // 			desc: "go only",
 // 			go_mod: `
 // module test
 
-// go 1.23.0
+// go 1.17
+
+// require (
+//     github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+// )
 // `,
 // 			want: "go1.23.0",
 // 		},
+		{
+			desc: "missing go",
+			go_mod: `
+module test
+
+require (
+    github.com/bazelbuild/rules_go v0.53.0  // unused, just here to test the go.mod parser
+)
+`,
+			want: "go1.17",
+		},
 	} {
 		t.Run(test.desc, func(t *testing.T) {
 			if err := ioutil.WriteFile("go.mod", []byte(test.go_mod), 0o666); err != nil {
@@ -89,9 +123,10 @@ toolchain go1.24.1
 			}
 			args := []string{
 				"test",
-				"//:version_test",
+				"--enable_bzlmod",
 				"--test_arg=-version=" + test.want,
 				"--test_output=all",
+				"//:version_test",
 			}
 			if err := bazel_testing.RunBazel(args...); err != nil {
 				t.Fatal(err)

--- a/tests/core/from_go_mod_file/from_go_mod_file_test.go
+++ b/tests/core/from_go_mod_file/from_go_mod_file_test.go
@@ -1,0 +1,104 @@
+// Copyright 2025 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package from_go_mod_file_test
+
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+)
+
+func TestMain(m *testing.M) {
+	bazel_testing.TestMain(m, bazel_testing.Args{
+		Main: `
+-- BUILD.bazel --
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+
+go_test(
+    name = "version_test",
+    srcs = ["version_test.go"],
+)
+
+-- version_test.go --
+package version_test
+
+import (
+    "fmt"
+    "flag"
+    "runtime"
+    "testing"
+)
+
+var want = flag.String("version", "", "")
+
+func Test(t *testing.T) {
+    fmt.Println(runtime.Version())
+    if v := runtime.Version(); v != *want {
+        t.Errorf("got version %q; want %q", v, *want)
+    }
+}
+`,
+		ModuleFileSuffix: `
+go_sdk = use_extension("@io_bazel_rules_go//go:extensions.bzl", "go_sdk")
+go_sdk.from_file(name = "go_sdk", go_mod = "//:go.mod")
+`,
+	})
+}
+
+func Test(t *testing.T) {
+	for _, test := range []struct{
+		desc, go_mod, want string
+	}{
+		{
+			desc: "toolchain",
+			go_mod: `
+module test
+
+go 1.23.0
+
+toolchain go1.24.1
+`,
+			want: "go1.24.1",
+		},
+		{
+			desc: "go only",
+			go_mod: `
+module test
+
+go 1.23.0
+`,
+			want: "go1.23.0",
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			if err := ioutil.WriteFile("go.mod", []byte(test.go_mod), 0o666); err != nil {
+				t.Fatal(err)
+			}
+			args := []string{
+				"test",
+				"//:version_test",
+				"--test_arg=-version=" + test.want,
+				"--test_output=all",
+			}
+			if err := bazel_testing.RunBazel(args...); err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+
+

